sycl/source/detail/pi.cpp

```diff
@@ -685,45 +685,49 @@ static uint16_t getELFHeaderType(const unsigned char *ImgData, size_t ImgSize) {
 sycl::detail::pi::PiDeviceBinaryType
 getBinaryImageFormat(const unsigned char *ImgData, size_t ImgSize) {
   // Top-level magic numbers for the recognized binary image formats.
-  struct {
-    sycl::detail::pi::PiDeviceBinaryType Fmt;
-    const uint32_t Magic;
-  } Fmts[] = {{PI_DEVICE_BINARY_TYPE_SPIRV, 0x07230203},
-              {PI_DEVICE_BINARY_TYPE_LLVMIR_BITCODE, 0xDEC04342},
-              // 'I', 'N', 'T', 'C' ; Intel native
-              {PI_DEVICE_BINARY_TYPE_NATIVE, 0x43544E49}};
-
-  if (ImgSize >= sizeof(Fmts[0].Magic)) {
-    std::remove_const_t<decltype(Fmts[0].Magic)> Hdr = 0;
-    std::copy(ImgData, ImgData + sizeof(Hdr), reinterpret_cast<char *>(&Hdr));
-
-    // Check headers for direct formats.
-    for (const auto &Fmt : Fmts) {
-      if (Hdr == Fmt.Magic)
-        return Fmt.Fmt;
-    }
+  auto MatchMagicNumber = [&](auto Number) {
+    if (ImgSize < sizeof(Number))
+      return false;
+    return std::memcmp(ImgData, &Number, sizeof(Number)) == 0;
```

Comment:
Nit, but I am not a huge fan of manual short-circuiting.
```suggestion
    return ImgSize >= sizeof(Number) && std::memcmp(ImgData, &Number, sizeof(Number)) == 0;
```